### PR TITLE
Add mem tracker to to trace rf and tbl desc

### PR DESCRIPTION
**Pull request by satanson**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Add query-level MemTracker in pipeline::QueryContext to support accounting on tbl_desc and runtime filters.

### CODE REVIEW

**Review by laotan332 — DISMISSED** [review]

File: be/src/exec/pipeline/query_context.cpp [hunk 1]
```diff
@@ -34,6 +31,49 @@ void QueryContext::cancel(const Status& status) {
...
+int64_t QueryContext::compute_query_mem_limit(int64_t parent_mem_limit, int64_t per_instance_mem_limit,
+                                              size_t pipeline_dop) {
+    // no mem_limit
+    if (per_instance_mem_limit == -1) {
+        return -1;
+    }
+    int64_t mem_limit = per_instance_mem_limit;
+    // query's mem_limit = per-instance mem_limit * num_instances * pipeline_dop
+    static constexpr int64_t MEM_LIMIT_MAX = std::numeric_limits<int64_t>::max();
+    if (MEM_LIMIT_MAX / total_fragments() < mem_limit) {
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/query_context.cpp:43]: why not check the overflow in one-line

File: be/src/runtime/exec_env.cpp [hunk 2]
```diff
@@ -233,6 +236,12 @@ const std::string& ExecEnv::token() const {
     return _master_info->token;
 }
 
+void ExecEnv::add_rf_event(const RfTracePoint& pt) {
+    std::string msg = strings::Substitute("$0($1)", std::move(pt.msg),
```
Comment (mofeiatwork) [inline be/src/runtime/exec_env.cpp:240]: why move it

File: be/src/runtime/runtime_filter_cache.h [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    const size_t _slot_mask;
+    std::vector<std::shared_mutex> _mutexes;
+    std::vector<std::unordered_map<TUniqueId, RfCacheValuePtr>> _filter_maps;
+    std::vector<std::shared_mutex> _event_mutexes;
+    std::vector<std::unordered_map<TUniqueId, RfEventValuePtr>> _event_maps;
+    mutable std::atomic<bool> _enable_trace{false};
+    std::atomic<size_t> _cache_times{0};
+    std::atomic<size_t> _use_times{0};
+    std::atomic<bool> _stop{false};
+    std::shared_ptr<std::thread> _clean_thread;
```
Comment (mofeiatwork) [inline be/src/runtime/runtime_filter_cache.h:79]: so many threads!!!

File: be/src/runtime/runtime_filter_cache.h [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    std::thread& clean_thread() { return *_clean_thread.get(); }
+
+private:
+    static void _clean_thread_func(RuntimeFilterCache* cache);
+    void _clean_events(bool force);
+    void _clean_filters();
+    size_t _slot_idx(const TUniqueId& query_id);
+    const size_t _num_slots;
+    const size_t _slot_mask;
+    std::vector<std::shared_mutex> _mutexes;
```
Comment (mofeiatwork) [inline be/src/runtime/runtime_filter_cache.h:71]: why not make a struct for mutex and map

File: be/src/runtime/runtime_filter_worker.cpp [hunk 5]
```diff
@@ -509,10 +597,138 @@ void RuntimeFilterWorker::_receive_total_runtime_filter(PTransmitRuntimeFilterPa
...
+    auto passthrough_delivery = params.data().size() <= config::deliver_broadcast_rf_passthrough_bytes_limit;
+    if (passthrough_delivery) {
+        _deliver_broadcast_runtime_filter_passthrough(std::move(params), std::move(destinations), timeout_ms);
+    } else {
+        _deliver_broadcast_runtime_filter_relay(std::move(params), std::move(destinations), timeout_ms);
+    }
+}
+
+void RuntimeFilterWorker::_deliver_broadcast_runtime_filter_relay(PTransmitRuntimeFilterParams&& request,
+                                                                  std::vector<TRuntimeFilterDestination>&& destinations,
```
Comment (mofeiatwork) [inline be/src/runtime/runtime_filter_worker.cpp:647]: why is the rvalue-ref necessary

File: be/src/runtime/runtime_filter_worker.cpp [hunk 6]
```diff
@@ -509,10 +597,138 @@ void RuntimeFilterWorker::_receive_total_runtime_filter(PTransmitRuntimeFilterPa
...
+            }
+            _exec_env->add_rf_event({request.query_id(), request.filter_id(), dest.address.hostname,
+                                     "DELIVER_BROADCAST_RF_PASSTHROUGH"});
+            send_rpc_runtime_filter(stub, rpc_closure, timeout_ms, request);
+        }
+
+        for (auto& rpc_closure : rpc_closures) {
+            brpc::Join(rpc_closure->cntl.call_id());
+            rpc_closure->unref();
+            delete rpc_closure;
```
Comment (mofeiatwork) [inline be/src/runtime/runtime_filter_worker.cpp:714]: will the `unref` release the object?

Comment (satanson) [reply be/src/runtime/runtime_filter_worker.cpp]: not, unref is matched with ref above

File: be/src/runtime/runtime_filter_worker.cpp (hunk 5 above)
Comment (satanson) [reply be/src/runtime/runtime_filter_worker.cpp]: move works well here

File: be/src/runtime/exec_env.cpp (hunk 2 above)
Comment (satanson) [reply be/src/runtime/exec_env.cpp]: pt.msg is not re-used after this point.

File: be/src/exec/pipeline/query_context.cpp [hunk 7]
```diff
@@ -34,6 +31,49 @@ void QueryContext::cancel(const Status& status) {
...
 
+int64_t QueryContext::compute_query_mem_limit(int64_t parent_mem_limit, int64_t per_instance_mem_limit,
+                                              size_t pipeline_dop) {
+    // no mem_limit
+    if (per_instance_mem_limit == -1) {
+        return -1;
+    }
+    int64_t mem_limit = per_instance_mem_limit;
+    // query's mem_limit = per-instance mem_limit * num_instances * pipeline_dop
+    static constexpr int64_t MEM_LIMIT_MAX = std::numeric_limits<int64_t>::max();
```
[the remark this answers is not in the document]
Comment (satanson) [reply be/src/exec/pipeline/query_context.cpp:42]: yes, read the whole function,  mem_limit not exceeds parent_mem_limit

File: be/src/exec/pipeline/query_context.cpp (hunk 1 above)
Comment (satanson) [reply be/src/exec/pipeline/query_context.cpp]: ok

**Review by murphyatwork — DISMISSED** [review]

**Review by kangkaisen — DISMISSED** [review]

File: be/src/exec/pipeline/fragment_executor.cpp [hunk 8]
```diff
@@ -158,19 +148,25 @@ Status FragmentExecutor::prepare(ExecEnv* exec_env, const TExecPlanFragmentParam
...
-        runtime_state->init_mem_trackers(query_id, nullptr);
+        auto* parent_mem_tracker = wg != nullptr ? wg->mem_tracker() : exec_env->query_pool_mem_tracker();
+        auto per_instance_mem_limit = query_options.__isset.mem_limit ? query_options.mem_limit : -1;
+        int64_t query_mem_limit = _query_ctx->compute_query_mem_limit(parent_mem_tracker->limit(),
+                                                                      per_instance_mem_limit, degree_of_parallelism);
+        _query_ctx->init_mem_tracker(query_mem_limit, parent_mem_tracker);
     }
+
+    auto query_mem_tracker = _query_ctx->mem_tracker();
+    MemTracker* prev_tracker = tls_thread_status.set_mem_tracker(query_mem_tracker.get());
```
Comment (trueeyu) [inline be/src/exec/pipeline/fragment_executor.cpp:163]: can use `SCOPED_THREAD_LOCAL_MEM_TRACKER_SETTER`

File: be/src/exec/pipeline/hashjoin/hash_join_build_operator.cpp [hunk 9]
```diff
@@ -64,10 +65,16 @@ Status HashJoinBuildOperator::set_finishing(RuntimeState* state) {
     auto& partial_in_filters = _join_builder->get_runtime_in_filters();
     auto& partial_bloom_filter_build_params = _join_builder->get_runtime_bloom_filter_build_params();
     auto& partial_bloom_filters = _join_builder->get_runtime_bloom_filters();
+
+    auto mem_tracker = state->query_ctx()->mem_tracker();
```
Comment (trueeyu) [inline be/src/exec/pipeline/hashjoin/hash_join_build_operator.cpp:69]: Is `set_finished` not counted in the query memory before?

Comment (satanson) [reply be/src/exec/pipeline/hashjoin/hash_join_build_operator.cpp:69]: counted in instance-level memory tracker

File: be/src/exec/pipeline/fragment_executor.cpp (hunk 8 above)
Comment (satanson) [reply be/src/exec/pipeline/fragment_executor.cpp]: ok

File: be/src/exec/pipeline/query_context.cpp [hunk 10]
```diff
@@ -35,6 +31,44 @@ void QueryContext::cancel(const Status& status) {
...
+    if (MEM_LIMIT_MAX / total_fragments() / pipeline_dop < mem_limit) {
+        mem_limit *= total_fragments() * pipeline_dop;
+    } else {
+        mem_limit = MEM_LIMIT_MAX;
+    }
+    // query's mem_limit never exceeds its parent's limit if it exists
+    return parent_mem_limit == -1 ? mem_limit : std::min(parent_mem_limit, mem_limit);
```
Comment (trueeyu) [inline be/src/exec/pipeline/query_context.cpp:49]: why compute std::min(parent_mem_limit, mem_limit) ?

**Review by murphyatwork — DISMISSED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by trueeyu — APPROVED** [review]

**Pull request merged**